File: src/main/java/com/wanted/repository/OrderRepositoryImpl.java
package com.wanted.repository;

import com.querydsl.core.types.dsl.Expressions;
import com.querydsl.jpa.impl.JPAQueryFactory;
import com.wanted.dto.PurchaseDto;
import com.wanted.dto.QPurchaseDto;
import jakarta.persistence.EntityManager;

import java.util.List;

import static com.wanted.entity.QMember.member;
import static com.wanted.entity.QOrder.order;
import static com.wanted.entity.QOrderProduct.orderProduct;
import static com.wanted.entity.QProduct.product;

public class OrderRepositoryImpl implements OrderRepositoryCustom{

    private final JPAQueryFactory queryFactory;

    public OrderRepositoryImpl(EntityManager em) {
        this.queryFactory = new JPAQueryFactory(em);
    }

    @Override
    public List<PurchaseDto> findAllPurchaseDtos() {
        return queryFactory
                .select(new QPurchaseDto(
                        order.id.as("order_id"),
                        member.email,
                        product.name,
                        orderProduct.count,
                        product.price,
                        Expressions.numberTemplate(Long.class, "{0} * {1}", orderProduct.count, product.price).as("totalPrice"),
                        order.status.stringValue().as("status")))
                .from(order)
                .join(order.member,member)
                .join(order.orderProducts, orderProduct)
                .join(orderProduct.product, product)
                .fetch();
    }
}

File: src/main/kotlin/wanted/market/api/order/repository/OrderRepositoryImpl.kt
package wanted.market.api.order.repository

import com.querydsl.jpa.impl.JPAQueryFactory
import org.springframework.stereotype.Repository
import wanted.market.api.order.domain.dto.out.RetrievePurchaseHistoryResult
import wanted.market.api.order.domain.dto.out.RetrieveReservationHistoryResult
import wanted.market.api.order.domain.entity.QOrder.order
import wanted.market.api.order.domain.entity.QOrderItem.orderItem

@Repository
class OrderRepositoryImpl(
    private val queryFactory: JPAQueryFactory
) : OrderRepositoryCustom {

    override fun findAllPurchaseHistory(productId: Long, memberId: Long) : List<RetrievePurchaseHistoryResult> {
        val results = queryFactory
            .select(order)
            .from(order)
            .leftJoin(order.orderItems, orderItem).fetchJoin()
            .where(order.buyer.id.eq(memberId), orderItem.product.id.eq(productId))
            .fetch()
        return results.map { RetrievePurchaseHistoryResult.from(it) }
    }

    override fun findAllReservationHistory(productId: Long, memberId: Long): List<RetrieveReservationHistoryResult> {
        val results = queryFactory
            .select(order)
            .from(order)
            .leftJoin(order.orderItems, orderItem).fetchJoin()
            .where(order.seller.id.eq(memberId), orderItem.product.id.eq(productId))
            .fetch()
        return results.map { RetrieveReservationHistoryResult.from(it) }
    }
}